src/ucs/datastruct/ptr_array.c

```diff
@@ -172,37 +190,128 @@ unsigned ucs_ptr_array_insert(ucs_ptr_array_t *ptr_array, void *value,
     }
 
     /* Get the first item on the free list */
-    elem_index = ptr_array->freelist;
-    ucs_assert(elem_index != UCS_PTR_ARRAY_SENTINEL);
-    elem = &ptr_array->start[elem_index];
+    element_index = ptr_array->freelist;
+    ucs_assert(element_index != UCS_PTR_ARRAY_SENTINEL);
+
+    elem = &ptr_array->start[element_index];
+
+    /* Remove from free list and populate */
+    ptr_array->freelist =
+        ucs_ptr_array_freelist_get_next_params(*elem, &size_free_ahead);
+    *elem               = (uintptr_t)value;
+
+    return element_index;
+}
+
+void ucs_ptr_array_remove(ucs_ptr_array_t *ptr_array, unsigned element_index)
+{
+    ucs_ptr_array_elem_t *elem = &ptr_array->start[element_index];
+    ucs_ptr_array_elem_t *next_elem;
+    uint32_t size_free_ahead;
+
+    ucs_assert_always(!ucs_ptr_array_is_free(ptr_array, element_index));
+
+    if (((element_index + 1) < ptr_array->size) &&
+        (__ucs_ptr_array_is_free(ptr_array->start[element_index + 1]))) {
+        next_elem = &ptr_array->start[element_index + 1];
+        ucs_ptr_array_freelist_get_next_params(*next_elem, &size_free_ahead);
```

Comment:
you can use `ucs_ptr_array_size_free_get` here, because you do not need next

Reply:
Actually, size_free_ahead is required here, see one line below:
    if (((element_index + 1) < ptr_array->size) &&
        (__ucs_ptr_array_is_free(ptr_array->start[element_index + 1]))) {
        next_elem = &ptr_array->start[element_index + 1];
        ucs_ptr_array_freelist_get_next_params(*next_elem, &size_free_ahead);
        size_free_ahead++;
    } else {
        size_free_ahead = 1;
    }

Reply:
my comment was about next, not the `size_free_ahead`